src/index.cpp

```diff
@@ -2022,14 +2023,10 @@ std::pair<uint32_t, uint32_t> Index<T, TagT, LabelT>::_search_with_filters(const
                                                                            float *distances)
 {
     auto converted_label = this->get_converted_label(raw_label);
-    if (typeid(uint64_t *) == indices.type())
```

Comment:
i think indices refers to actual location of vectors, not the tags. it is used for a bulk index (not a streaming scenario)  where we implicitly assume 0th vector has index= 0, 1st vector has index 1, etc. In some ads use case, they wanted the index itself to be either uint64 or uint32, hence this type casting. This is different from tags, so please leave this if condition in place.

Reply:
we use the same search function to support static index and streaming index:
https://github.com/microsoft/DiskANN/blob/main/src/index.cpp#L2103

That means the tag type must support to convert to IdType even we don't go through the streaming part as the template instantiate. And as the tag may not be a basic type(uint32, uint64, etc), we can't instantiate the template like:

Index<T=uint8, TagT=tag_uint128, LabelT=uint32, IdType=uint64> because the following code will compile error (line 2100):
TagT tag;
if (_location_to_tag.try_get(best_L_nodes[i].id, tag))
{
    indices[pos] = (IdType)tag;
}

The code I removed here is trying to convert IdType to uint32 or uint64, that will trigger a template instantiate for all index template define in line 3327 including Index<T=uint8, TagT=tag_uint128, LabelT=uint32>:
https://github.com/microsoft/DiskANN/blob/main/src/index.cpp#L3327

Reply:
The code is already use force cast to convert Tag to IdType, so I think it should be safe here, and I also think define uint64 tag but want to return uint32 indices doesn't make sense.

The TagT is uint32 by default, and the indices is uint32 or uint64, so nothing is broken to remove this code.

If we still need to keep this code, we have to separate the searching funtion of static index and streaming index, like:
search_with_filters_tags();

Reply:
I think separating is better. Ads team wanted IndexType = uint32 or uint64 for static index. So the IndexType is used for static build, and TagT is used for dynamic index, and there is bug in existing code itself even before your PR. 

To fix this, can you just add "bool use_filters = false, const std::string raw_label="" as default arguments in all variants of search_with_tags (with and without _ and in abstract_index and index) to enable filter support in dynamic indices? If you want I can draft a PR for this and you can rebase. Let me know if you want to do this or me.

Reply:
on discussing with Gopal, I think it is best to go ahead with this PR and fix that issue in next PR.

Reply:
Change search_with_tags() to support filtered search, and remove the code of streaming in search_with_filters().